src/Trace/Sampler/QpsSampler.php

```
@@ -0,0 +1,145 @@
+     * @param array $options [optional] {
+     *     configuration options.
+     *
+     *     @type string $cacheItemClass The class of the item to use. This class must implement
+     *           CacheItemInterface.
+     *     @type float $rate The number of queries per second to allow. Must be less than or equal to 1.
+     *           **Defaults to** `0.1`
+     *     @type string $key The cache key to use. **Defaults to** `__google_cloud_trace__`
+     * }
+     */
+    public function __construct(CacheItemPoolInterface $cache = null, $options = [])
+    {
+        $this->cache = $cache ?: $this->defaultCache();
+        if (!$this->cache) {
+            throw new \InvalidArgumentException('Cannot use QpsSampler without providing a PSR-6 $cache');
+        }
+
+        $options += [
+            'cacheItemClass' => self::DEFAULT_CACHE_ITEM_CLASS,
+            'rate' => self::DEFAULT_QPS_RATE,
+            'key' => self::DEFAULT_CACHE_KEY
+        ];
+        $this->cacheItemClass = $options['cacheItemClass'];
+        $this->rate = $options['rate'];
+        $this->key = $options['key'];
+
+        if ($this->rate > 1 || $this->rate <= 0) {
+            throw new \InvalidArgumentException('QPS sampling rate must be less that 1 query per second');
+        }
+    }
+
+    /**
+     * Returns whether or not the request should be sampled.
+     *
+     * @return bool
+     */
+    public function shouldSample()
+    {
+        // We will store the microtime timestamp in the cache because some
+        // cache implementations will not let you use expiry for anything less
+        // than 1 minute
+        if ($item = $this->cache->getItem($this->key)) {
+            if ((float) $item->get() > microtime(true)) {
+                return false;
+            }
+        }
+
+        $item = new $this->cacheItemClass($this->key);
+        $item->set(microtime(true) + 1.0 / $this->rate);
+
+        // TODO: what if the cache fails to save?
```

I think this is actually open for discussion/change. It's currently set to silently fail if the cache implementation fails. `save()` should return `false` on failure (after catching any driver's exception). There's no standardized way to get at the exception from PSR-6.

We could throw our own exception here if the cache write fails, but the error wouldn't have much information on it besides cache writing failed. If we do silently fail, the side effect is that we end up tracing all requests (which I'm ok with).